Public/Src/Engine/Processes/SharedOpaqueOutputLogger.cs

```diff
@@ -0,0 +1,233 @@
+        public SharedOpaqueOutputLogger(string logPath, [CanBeNull] IReadOnlyList<string> rootDirectories)
+        {
+            LogPath = logPath;
+            RootDirectories = rootDirectories;
+            m_recordedPathsCache = new HashSet<AbsolutePath>();
+
+            m_lazyBxlWriter = Lazy.Create(() =>
+            {
+                Directory.CreateDirectory(Path.GetDirectoryName(LogPath));
+                return new BuildXLWriter(
+                    stream: new FileStream(LogPath, FileMode.Create, FileAccess.Write, FileShare.Read | FileShare.Delete),
+                    debug: false,
+                    logStats: false,
+                    leaveOpen: false);
+            });
+        }
+
+        private const string LogFilePrefix = "Pip";
+        private const string LogFileSuffix = ".outlog";
+
+        /// <summary>
+        /// Given a root directory (<paramref name="searchRootDirectory"/>), returns the full path to the log file corresponding to process <paramref name="process"/>.
+        /// </summary>
+        public static string GetOutputLogFileForProcess(PathTable pathTable, AbsolutePath searchRootDirectory, Process process)
+        {
+            Contract.Requires(searchRootDirectory.IsValid);
+
+            var semiStableHashX16 = string.Format(CultureInfo.InvariantCulture, "{0:X16}", process.SemiStableHash);
+            var subDirName = semiStableHashX16.Substring(0, 3);
+
+            return searchRootDirectory.Combine(pathTable, subDirName).Combine(pathTable, $"{LogFilePrefix}{semiStableHashX16}{LogFileSuffix}").ToString(pathTable);
+        }
+
+        /// <summary>
+        /// Finds and returns all output log files that exist in directory denoted by <paramref name="directory"/>
+        /// </summary>
+        /// <remarks>
+        /// CODESYNC: must be consistent with <see cref="GetOutputLogFileForProcess(PathTable, AbsolutePath, Process)"/>
+        /// </remarks>
+        public static string[] FindAllProcessPipOutputLogFiles(string directory)
+        {
+            return Directory.Exists(directory)
+                ? Directory.EnumerateFiles(directory, $"{LogFilePrefix}*{LogFileSuffix}", SearchOption.AllDirectories).ToArray()
+                : CollectionUtilities.EmptyArray<string>();
+        }
+
+        /// <summary>
+        /// Returns all paths recorded in the <paramref name="filePath"/> file.
+        /// </summary>
+        /// <remarks>
+        /// Those paths are expected to be absolute paths of files/directories that were written to by the previous build.
+        ///
+        /// NOTE: this method does not validate the recorded paths in any way.  That means that each returned string may be
+        ///   - a path pointing to an absent file
+        ///   - a path pointing to a file
+        ///   - a path pointing to a directory.
+        /// 
+        /// NOTE: unless the log file was corrupted (i.e., it was produced by an instance of this class):

```

Comment:
Should this say `wasn't` because only other impls could corrupt the file?

Reply:
Rephrased it.
```suggestion
        /// NOTE: if the log file was produced by an instance of this class (and wasn't corrupted in any way)
```